docs/rules/use-object-in-propequal.md

```diff
@@ -0,0 +1,39 @@
+# Enforce use of objects as expected value in `assert.propEqual` (use-object-in-propequal)
+
+The `assert.propEqual` assertion is for the strict-equality comparison of own-properties
+of two objects. If the expected value is a string or other non-object, the assertion
+result can be unexpected.
+
+For string comparisons, `assert.strictEqual` should be used. For arrays,
+`assert.deepStrictEqual` should be used.
+
+## Rule Details
+
+The following patterns are considered warnings:
+
+```js
+assert.propEqual(actual, 0);
+
+assert.propEqual(actual, "foo");
+
+assert.propEqual(actual, ["array"]);
```

Comment:
Arrays might be alright actually, I can see that being useful. E.g. similar to how one can use an object literal to verify the own properties of any class instance (ignoring inheritance), one can also use an array to verify the sequential own properties of other array-like objects.

I've not used it that way myself, but it stood out to me as I generally think of arrays as being an object, not a primitive or non-object. Having said that, regexes and functions are of course fully objects and should receive warnings here.

Reply:
Thanks for reviewing!

My thinking was that deepEqual or deepStrictEqual would be better for arrays.

As a QUnit maintainer, what do you think? Is there a reason to use propEqual over deepEqual?

Reply:
If the `actual` value is also a plain array, then `propEqual()` indeed has no reason for being used as it would behave identically to `deepEqual()`. The main reason for `propEqual()` is for when the `actual` value is more complex, e.g. a class instance or other prototypal oject with inherited properties or methods. In that case you can use propEqual to only compare the own properties. Perhaps because the class is an internal implementation detail that doesn't matter to the API, or for other convenience reasons you don't wish to setup or assert the inheritence.

With that in mind, whilst I haven't come across it personaly, I can see an array being used in that same manner. Just as class instances are essentially plain objects with some extra inherited stuff, arrays too can be subclassed and/or mimicked by complex objects.

For example:

```js
// Modern
class MyList extends Array {}

// Classic or functional
function MyList () {}
MyList.prototype = Object.create(Array.prototype);
MyList.prototype.constructor = MyList;

x = new MyList();
x.push("foo");
// Object MyList { 0: "foo", length: 1 }
```

If its constructor is private or otherwise non-trivial, and `[0]`, `[1]`, `[2]` etc, are its only public properties, then `propEqual(x, ["foo"])` can serve as a convenient way to strict assert the list contents in full.

It is possible that if we see this in practice with a realistic use case, we might judge it as poor or overly-clever. But I'm thinking we may want to reserve judgement to avoid avoidable nuissance to users, until/unless we know of a case where it was definitely unintended or fragile. No strong opinions either way 🙂

Reply:
It sounds like maybe an option would be a good idea, so users could opt in and allow array expressions if they know it would be useful, but by default the rule would warn against it. Do you think that could be a good approach?

Can you think of other object-ish types that could benefit from this, besides arrays?

Reply:
I'm not aware of other ones, no. I do still lean towards allowing it by default, especially if we plan to add it to the recommended preset in the future. Use with arrays is imho harmless (no downsides afaik), and we know at least of plausible ways in which it benefits.

To me, the main problem with accidental use of propEqual is to avoid false positives (e.g. when boxing a number or boolean in an object, that object has no own properties so it would likely pass even if the value is wrong), and to avoid confusing diffs (e.g. a string showing up as it it were an object, which the user did not enter so the diff can be rather strange). For arrays it behaves identical to deepEqual both functionally and in terms of developer experience.

Reply:
(Note to self: We might want to throw in propEqual for the definitely-wrong case of passing a primitive value, perhaps something for QUnit 3).

Reply:
Okay, I'll allow arrays by default. Users can use `no-restricted-syntax` if they want to catch the array expressions case.